numpy/lib/arraysetops.py

```diff
@@ -593,6 +615,75 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     # Ensure that iteration through object arrays yields size-1 arrays
     if ar2.dtype == object:
         ar2 = ar2.reshape(-1, 1)
+    # Convert booleans to uint8 so we can use the fast integer algorithm
+    if ar1.dtype == bool:
+        ar1 = ar1.view(np.uint8)
+    if ar2.dtype == bool:
+        ar2 = ar2.view(np.uint8)
+
+    # Check if we can use a fast integer algorithm:
+    integer_arrays = (np.issubdtype(ar1.dtype, np.integer) and
+                      np.issubdtype(ar2.dtype, np.integer))
+
+    if kind not in {None, 'mergesort', 'dictionary'}:
+        raise ValueError(
+            "Invalid kind: {0}. ".format(kind)
+            + "Please use None, 'mergesort' or 'dictionary'.")
+
+    if integer_arrays and kind in {None, 'dictionary'}:
+        ar2_min = np.min(ar2)
+        ar2_max = np.max(ar2)
+        ar1_size = ar1.size
+        ar2_size = ar2.size
+
+        # Check for integer overflow
+        with np.errstate(over='raise'):
+            try:
+                ar2_range = ar2_max - ar2_min
+
+                # Optimal performance is for approximately
+                # log10(size) > (log10(range) - 2.27) / 0.927.
+                # However, here we set the requirement that
+                # the intermediate array can only be 6x
+                # the combined memory allocation of the original
+                # arrays.
+                # (see discussion on 
+                # https://github.com/numpy/numpy/pull/12065)
+                below_memory_constraint = (
+                    ar2_range <= 6 * (ar1_size + ar2_size)
+                )
+            except FloatingPointError:
+                below_memory_constraint = False
```

Comment:
Just a thought:  Maybe move this up and add a `finally:` clause?  That would make it a bit more clear where the overflow would occur.

Reply:
I suppose this needs an extra error path?  If the method is forced, but is not compatible with the input range?

Reply:
Which code do you propose to move to the `finally:` clause?

Reply:
I added a `range_safe_from_overflow` parameter that will forcefully skip the `"dictionary"` method if there's an overflow. Should I add an error if it gets skipped despite explicitly setting it, or just silently do so?

Reply:
The `ar2_range <= 6 * (ar1_size + ar2_size)` calculation has no business of being guarded? Although you should (soon must maybe) make it `int(ar2_range)`.
Since you renamed the first variable, it doesn't even need to be in a `finally` clause.

Maybe someone else has an opinion.  I might just go with an error (you can still remove it). If  at all an option, it should be part of the kind like `"try-dictionary"`...

Reply:
Good point, will move it outside.

Reply:
Would something like
```python
        range_safe_from_overflow = ar2_range < np.iinfo(ar2.dtype).max
```
be what you had in mind? Then I can use:
```python
        ar2_range = int(ar2_max) - int(ar2_min)
```
which is cleaner as I don't have to use the `errstate`.

Reply:
Implemented this. Let me know what you think. I'll also add the requested unittests, etc.

Reply:
Unittests implemented. It should now test every error:
- `kind` is not of the three available.
- non-integral array passed but kind=dictionary.
- integer overflow but kind=dictionary.

It also tests lack of errors:
- integer overflow, but doesn't trigger error as kind=None or kind=mergesort.